addons/website_sale_stock/controllers/main.py

```diff
@@ -26,3 +27,23 @@ def shop_payment_transaction(self, *args, **kwargs):
         if values:
             raise ValidationError(' '.join(values))
         return super().shop_payment_transaction(*args, **kwargs)
+
+
+class WebsiteSale(website_sale_controller.WebsiteSale):
+    @http.route(['/shop/add/stock_notification'], type="json", auth="public", website=True)
+    def add_stock_email_notification(self, email, product_id):
+        if not email_re.match(email):
+            return {'error': BadRequest("Invalid Email")}
+
+        product = request.env['product.product'].browse(int(product_id))
+        partners = request.env['res.partner'].sudo()._mail_find_partner_from_emails([email], force_create=True)
+        partner = partners[0]
+
+        if not product.has_stock_notification(partner):
+            product.sudo().stock_notification_partner_ids += partner
+
+    def _prepare_product_values(self, product, category='', search='', **kwargs):
+        values = super()._prepare_product_values(product, category, search, **kwargs)
+        # We need the user mail to prefill the back of stock notification, so we put it in the value that will be sent
+        values['email'] = request.env.user.email
```

Comment:
I think it would be beneficial to name this more explicitly, something like `user_email` would be more appropriate.